// sourcemap-blender/dev/dev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const devUtils_1 = require("./devUtils");
const escodegen = require("escodegen");
const Blender_1 = require("../Blender");
const fs = require("fs");
const path = require("path");
const cnt = fs.readFileSync(path.resolve(__dirname, "../src/dev/test.txt")).toString();
const simpleString = cnt;
const output = devUtils_1.transpileCode(simpleString);
let modifiedCode = escodegen.generate(output.ast);
modifiedCode = modifiedCode.replace("'use strict';", '');
//console.log(modifiedCode);
// modifiedCode = `
// window.somestuff = [1,2, "3", "oho"]
// console.log("some bullshit here")
// ${modifiedCode}
// `;
const blender = new Blender_1.Blender({
    parse: devUtils_1.acornParse,
    originalMap: output.map,
    originalAST: output.ast,
    modifiedCode: modifiedCode
});
const data = blender.blend({ sourceMappingURL: true });
data.then(data => {
    console.log(data.code);
    devUtils_1.writeTestIndex({
        source: data.code,
        maps: data.map //data.map
    });
    // writeTestIndex({
    // 	source: data.code,
    // 	maps: data.map
    // });
    // writeTestIndex({
    //     source :output.transpiled,
    //     maps : JSON.stringify(output.map)
    // })
});
