// sourcemap-blender/dev/devUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const acorn = require("acorn");
const ts = require("typescript");
const path = require("path");
const fs = require("fs");
const tsconfig = {
    compilerOptions: {
        module: "commonjs",
        target: "es5",
        sourceMap: true,
        inlineSources: true,
        //importHelpers: true,
        lib: ["es2015"],
        emitDecoratorMetadata: true,
        experimentalDecorators: true
    }
};
function writeTestIndex(input) {
    const content = `
    <!doctype html>
<html>
<head>
    <title></title>
</head>

<body>
		<script>window.exports = {}</script>
    <script type="text/javascript" src="/module.js"></script>
</body>

</html>`;
    const moduleJS = path.resolve("./dist/module.js");
    const sourceMaps = path.resolve("./dist/module.js.map");
    const index = path.resolve("./dist/index.html");
    fs.writeFileSync(moduleJS, input.source);
    fs.writeFileSync(sourceMaps, input.maps);
    fs.writeFileSync(index, content);
}
exports.writeTestIndex = writeTestIndex;
function writeToDist(name, contents) {
    const target = path.resolve("./dist/", name);
    fs.writeFileSync(target, contents);
}
exports.writeToDist = writeToDist;
function acornParse(source) {
    return acorn.parse(source, {
        ...{
            sourceType: "module",
            tolerant: true,
            locations: true,
            ranges: true,
            ecmaVersion: "2018"
        }
    });
}
exports.acornParse = acornParse;
function transpileCode(contents) {
    const response = ts.transpileModule(contents, tsconfig);
    const text = response.outputText;
    return {
        map: JSON.parse(response.sourceMapText),
        ast: acornParse(response.outputText),
        transpiled: response.outputText
    };
}
exports.transpileCode = transpileCode;
