// sourcemap-blender/test/testUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const devUtils_1 = require("../dev/devUtils");
const Blender_1 = require("../Blender");
const fuse_test_runner_1 = require("fuse-test-runner");
class TestBlender {
    constructor(opts) {
        this.opts = opts;
    }
    static async init(opts) {
        const test = new TestBlender(opts);
        const response = await test.run();
        return test;
    }
    shouldMatchMapping(mapping) {
        const map = JSON.parse(this.response);
        fuse_test_runner_1.should(map.mappings).equal(mapping);
    }
    async run() {
        const output = devUtils_1.transpileCode(this.opts.originalCode);
        const modifiedCode = this.opts.modify(output.transpiled);
        const blender = new Blender_1.Blender({
            parse: devUtils_1.acornParse,
            originalMap: output.map,
            originalAST: output.ast,
            modifiedCode: modifiedCode
        });
        const data = await blender.blend();
        this.response = data.map;
        return data.map;
    }
}
exports.TestBlender = TestBlender;
